data_sync/mvds.md

```diff
@@ -0,0 +1,130 @@
+# Minimum Viable Data Synchronization
+
+**DRAFT VERSION 0.5**
+
+*Written by Oskar Thorén oskar@status.im & Dean Eigenmann dean@status.im*
+
+## Table of Contents
+
+1. [Abstract](#abstract)
+2. [Definitions](#definitions)
+3. [Wire Protocol](#wire-protocol)
+    1. [Secure Transport](#secure-transport) 
+    2. [Payloads](#payloads)
+4. [Synchronization](#synchronization)
+    1. [State](#state)
+    2. [Flow](#flow)
+    3. [Retransmission](#retransmission)
+5. [Footnotes](#footnotes)
+
+## Abstract
+
+In this specification we describe a minimum viable protocol for data synchronization inspired by the Bramble Synchronization Protocol<sup>1</sup>. This protocol is designed to ensure reliable messaging between peers communicating across an unreliable peer-to-peer (P2P) network where they may be unreachable or unresponsive.
+
+We present a functional specification for future implementation<sup>2</sup> as well as reference simulation data which demonstrates its performance.
+
+## Definitions
+
+| Term | Description |
+|------|-------------|
+| **Peer** | The other nodes that a node is connected to. |
+| **Record** | Defines a payload element of either the type `OFFER`, `REQUEST`, `MESSAGE` or `ACK` |
+
+## Wire Protocol
+
+### Secure Transport
+
+This specification does not define anything related to the transport of packets. It is assumed that this is abstracted in such a way that any secure transport protocol could be easily implemented. Likewise, properties such as confidentiality, integrity, authenticity and forward secrecy are assumped to be provided by a layer below.
+
+### Payloads
+
+Payloads are implemented using [protocol buffers](https://developers.google.com/protocol-buffers/).
+
+```protobuf
+message Payload {
+  Ack ack = 1;
```

Comment:
Sounds like you want to use one_of here. 

```protobuf
message Payload {
   oneof msg_type {
      Ack ack = 1;
      Offer offer = 2;
      Request request = 3;
   }
   repeated Message messages = 4;
}
```

OR some enum

```protobuf
message Payload {
   enum Type {
      UNKNOWN = 0; // Default case, client should handle/reject a message with unspecified type.
      ACK = 1;
      OFFER = 2;
      REQUEST = 3;
   }
   uint64 id = 1; // Change this to whatever your identifier is.
   Type type = 2;
   repeated Message messages = 3;
}
```

Reply:
No, we don't wanna use one of. because a payload can contain all those `record`s.